lib/Dialect/HW/HWInstanceGraph.cpp

```diff
@@ -0,0 +1,34 @@
+//===- HWInstanceGraph.cpp - Instance Graph ---------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/HW/HWInstanceGraph.h"
+
+using namespace circt;
+using namespace hw;
+
+InstanceGraph::InstanceGraph(Operation *operation)
+    : InstanceGraphBase(operation) {
+  for (auto &node : nodes)
+    if (node.getModule().isPublic())
+      entry.addInstance({}, &node);
+}
+
+InstanceGraphNode *InstanceGraph::addModule(HWModuleLike module) {
+  auto *node = InstanceGraphBase::addModule(module);
+  if (module.isPublic())
+    entry.addInstance({}, node);
+  return node;
+}
```

Comment:
This was somewhat surprising behavior to me. Is there a good reason to create "dead", null instances? Are you doing this to block module erasure by guaranteeing that public modules always are always instantiated?

It was _very_ surprising to me that iterating over node instances could fail, i.e., `getInstance()` now has to be checked (which isn't necessary in FIRRTL's instance graph).

Is there a better way to do this, e.g., adding a bool state to the node to indicate public/private and using that to guard "hasNoUses"?

WDYT?